### DISCUSSION TITLE
New benchmark runner: Afterframe

### DISCUSSION THREAD

**Discussion by krausest** [krausest/js-framework-benchmark#1254] [Announcements]
So far the CPU benchmark measured duration by looking at the chrome timeline traces (the events one can see in the performance tab). This restricted testing to chrome only with the drivers: webdriver (selenium), puppeteer, playwright and webdriver-cdp. 

If we want to measure across browsers we have to measure on the client side, since safari and firefox seem to have no api to access something similar like timeline tracing. We had some functionality in the benchmarks that tried to do client side measuring for the duration in the past: (the commit for its removal was https://github.com/krausest/js-framework-benchmark/commit/89b546fe1b772e0f9f0b94da81b83513b6759e4b).
https://github.com/andrewiggins/afterframe takes a similar approach to get a callback after the rendering finished by using aMessageChannel and RAF.

It would be interesting to add such a benchmark driver and see if we can achieve cross browser testing.
The two classes to implement that are https://github.com/krausest/js-framework-benchmark/blob/master/webdriver-ts/src/benchmarksWebdriverAfterframe.ts and https://github.com/krausest/js-framework-benchmark/blob/master/webdriver-ts/src/forkedBenchmarkRunnerWebdriverAfterframe.ts. The test driver injects the afterframe code into the loaded page to compute the duration for the operation. This is similar to the other benchmark drivers where the whole duration from the click to finishing the repainting is measured.

**Comment by krausest** [comment]
The first question is: **How close are the CPU results for chrome and webdriver and webdriver-afterframe?**

When measuring with afterframe I haven't found any cross browser way to add a CPU slowdown, so all benchmarks were ran with full CPU speed.  I took three frameworks and a closer look at their results:
<img width="417" alt="Bildschirmfoto 2023-04-28 um 6 26 39 PM" src="https://user-images.githubusercontent.com/3374055/235202743-5532b7fa-0cf3-4e87-a90e-28314c817865.png">

As I hoped results are quite close for longer running benchmarks. Benchmarks below 1 FPS can vary more which was what I expected or feared. For almost all results the result from the chrome timeline was smaller that that from afterframe wich makes sense (afterframe should add some latency until the raf is executed).

Here's a chart showing the ratio of the duration reported by afterframe to the values reported by webdriver:
<img width="553" alt="Bildschirmfoto 2023-04-28 um 6 33 42 PM" src="https://user-images.githubusercontent.com/3374055/235204037-5fc24b07-c9e4-4fa7-98a3-293a31fae874.png">
The three large outliers are the select row benchmarks.
Thus I suggest to skip select row for benchmark testing with afterframe.

Without select row the chart looks smoother:
<img width="552" alt="Bildschirmfoto 2023-04-28 um 7 08 44 PM" src="https://user-images.githubusercontent.com/3374055/235210669-61fc79e1-ad3f-4368-9bcd-e581d4657027.png">

What's interesting is that variance for afterframe isn't much higher that for webdriver.

**Reply by GermanJablo** [reply]
Hi @krausest, thank you for opening this thread and for the incredible work you do with this benchmark.

Some questions:

1. Are the results of your image from the same iteration? If the answer is yes, how in the case of 14 and 24 afterframe could it be faster than timeline? (using the last image as reference)
2. You said:

> Chrome automation tools like puppeteer, playwright and webdriver can log the same events that the performance tab uses for its display

Could you explain to me how to access that information? (preferably with Playwright which is what I have the most experience with).

___ 

I would like to experiment to find a solution to this. Several options are discussed in [this very interesting blog article](https://nolanlawson.com/2018/09/25/accurately-measuring-layout-on-the-web/).

As you noticed, afterframe uses `rAF + MessageChannel`. I think there are other options to explore such as:
- PerformanceObserver
- transitionEnd and animationEnd events
- MutationObserver
- Web workers
